### PR TITLE
[Tests]: Unit Test for pkg/utils/protocols

### LINKED ISSUE #3835: [Feat:Improve Unit Test Coverage wherever required ]
https://mentorship.lfx.linuxfoundation.org/project/d3a1a899-1ca0-4e10-a402-01ef6fde26f2

---

# Improve Unit Test Coverage for Inspektor Gadget

## Current Situation

Inspektor Gadget currently has limited unit test coverage. Some critical code paths and edge cases remain untested, which could lead to undetected bugs and reduced code reliability.

## Impact

The lack of comprehensive unit tests in Inspektor Gadget can result in:  
1. Increased risk of regressions during future code changes.  
2. Difficulty in identifying and fixing bugs efficiently.  
3. Reduced confidence in the stability and reliability of the codebase.  

## Ideal Future Situation

Inspektor Gadget should have robust and comprehensive unit tests that cover all critical code paths and edge cases. This would:  
1. Increase the test coverage as much as possible.  
2. Ensure that the codebase is stable and resilient to changes.  
3. Facilitate faster development and debugging by identifying issues early.  

## Implementation Options

1. **Audit Existing Tests**: Identify gaps in the current test coverage and prioritize adding tests for untested or under-tested code.  
2. **Add Unit Tests**: Write detailed unit tests for Inspektor Gadget, covering:  
   - Edge cases.  
   - Common use cases.  
   - Potential failure scenarios.  
3. **Use Coverage Tools**: Utilize tools like `go test -cover` to measure coverage and identify areas that need improvement.  

## Additional Information

- Testing should adhere to the existing project standards and use consistent naming conventions.  
- Mocking and stubbing may be used where necessary to isolate unit tests from external dependencies.  
- Ensure that the added tests are maintainable and easy to understand.

### Priority 1
- [ ] [gadgets/trace_capabilities](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/trace_capabilities)
- [ ] [gadgets/trace_bind](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/trace_bind)
- [ ] [gadgets/snapshot_socket](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/snapshot_socket)
- [ ] [gadgets/trace_exec](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/trace_exec)
- [ ] [gadgets/trace_dns](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/gadgets/trace_dns)
- [ ] [pkg/operators](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/operators)
- [ ] [pkg/oci](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/oci)
- [ ] [pkg/datasource w/o compat](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/datasource)

### Priority 2
- [ ] [pkg/container-collection](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/container-collection)
- [ ] [pkg/types](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/types)
- [ ] [pkg/gadget-manifest](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/gadget-manifest)
- [ ] [pkg/container-utils/docker](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/container-utils/docker)

### Priority 3
- [ ] [pkg/uprobetracer/ldcache_parser.go](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/uprobetracer/ldcache_parser.go)
- [ ] [pkg/uprobetracer/usdt.go](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/uprobetracer/usdt.go)
- [ ] [pkg/gadget-service](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/gadget-service) (At least look into it)

### Priority 4
- [ ] [pkg/kallsyms](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/kallsyms)
- [ ] [pkg/logger](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/logger)
- [ ] [pkg/gadget-context](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/gadget-context)
- [ ] [pkg/btfhelpers](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/btfhelpers)
- [ ] [pkg/cachedmap](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/cachedmap)

### Priority 5
- Everything :)

### PR DESCRIPTION
**Pull request by upsaurav12**
# [Test]: Unit Test of pkg/utils/protocols
Fixes #3835
This Test ``GetProtocolNameFromNumber``. It has Two cases :- 
1. If number was found in ``protoNumberToName = map[int]string``. It will give true and Name of protocol
2.  If number was found in ``protoNumberToName = map[int]string``. It will give false and empty array

## How to use
Run
```
go test  ./... -v
```

## Testing done
```
=== RUN   TestGetProtocolNameByNumber
=== RUN   TestGetProtocolNameByNumber/number_was_found
=== RUN   TestGetProtocolNameByNumber/number_was_not_found
--- PASS: TestGetProtocolNameByNumber (0.00s)
    --- PASS: TestGetProtocolNameByNumber/number_was_found (0.00s)
    --- PASS: TestGetProtocolNameByNumber/number_was_not_found (0.00s)
PASS
ok      github.com/inspektor-gadget/inspektor-gadget/pkg/utils/protocols        0.002s
```

[Describe the testing you have done before submitting this PR. Please include both the commands you issued as well as the output you got.]

### CODE REVIEW

**Comment by upsaurav12** [comment]
@eiffel-fl PTAL

File: pkg/utils/protocols/protocols_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+func TestGetProtocolNameByNumber(t *testing.T) {
+	protocols := []struct {
+		description string
+		number      int
+		name        string
+		ok          bool
+	}{
+		{
```
Comment (burak-ok) [inline pkg/utils/protocols/protocols_test.go:29]: Would add another test with `-1` and maybe `TCP` since `TCP` will be the most used one for Inspektor Gadget

**Review by burak-ok — APPROVED** [review]
Looks good now. Thanks for your contribution
